server/src/main/java/io/druid/client/HttpServerInventoryView.java

```diff
@@ -0,0 +1,719 @@
+                {
+                  return input.segmentRemoved(druidServer.getMetadata(), segment);
+                }
+              }
+          );
+        }
+      };
+    }
+
+    //wait for first fetch of segment listing from server.
+    void awaitInitialization()
+    {
+      try {
+        if (!initializationLatch.await(serverHttpTimeout, TimeUnit.MILLISECONDS)) {
+          log.warn("Await initialization timed out for server [%s].", druidServer.getName());
+        }
+      } catch (InterruptedException ex) {
+        log.warn("Await initialization interrupted while waiting on server [%s].", druidServer.getName());
+        Thread.currentThread().interrupt();
+      }
+    }
+
+    DruidServerHolder updatedHolder(DruidServer server)
+    {
+      synchronized (lock) {
+        return new DruidServerHolder(server.addDataSegments(druidServer), counter) ;
+      }
+    }
+
+    Future<?> updateSegmentsListAsync()
+    {
+      try {
+        final String req;
+        if (counter != null) {
+          req = String.format(
+              "/druid-internal/v1/segments?counter=%s&hash=%s&timeout=%s",
+              counter.getCounter(),
+              counter.getHash(),
+              config.getServerTimeout()
+          );
+        } else {
+          req = String.format(
+              "/druid-internal/v1/segments?counter=-1&timeout=%s",
+              config.getServerTimeout()
+          );
+        }
+        URL url = new URL("http", serverHostAndPort.getHostText(), serverHostAndPort.getPort(), req);
+
+        final AtomicInteger returnCode = new AtomicInteger();
+        final AtomicReference<String> reasonString = new AtomicReference<>();
+
+        SequenceInputStreamResponseHandler responseHandler = new SequenceInputStreamResponseHandler()
+        {
+          @Override
+          public ClientResponse<InputStream> handleResponse(HttpResponse response)
+          {
+            returnCode.set(response.getStatus().getCode());
+            reasonString.set(response.getStatus().getReasonPhrase());
+            return super.handleResponse(response);
+          }
+        };
+
+        if (log.isDebugEnabled()) {
+          log.debug("Sending segment list fetch request to [%s] on URL [%s]", druidServer.getName(), url);
+        }
+
+        ListenableFuture<InputStream> future = httpClient.go(
+            new Request(HttpMethod.GET, url)
+                .addHeader(
+                    HttpHeaders.Names.ACCEPT,
+                    SmileMediaTypes.APPLICATION_JACKSON_SMILE
+                )
+                .addHeader(HttpHeaders.Names.CONTENT_TYPE, SmileMediaTypes.APPLICATION_JACKSON_SMILE),
+            responseHandler,
+            new Duration(serverHttpTimeout)
+        );
+
+        if (log.isDebugEnabled()) {
+          log.debug("Sent segment list fetch request to [%s]", druidServer.getName());
+        }
+
+        Futures.addCallback(
+            future,
+            new FutureCallback<InputStream>()
```

Comment:
I'm actually not sure which thread this callback will be called on, but I think that it might actually block in some not good locations (i.e. the Future will be done and returned when the first response packet comes back, but will block waiting on any subsequent chunks).  Rather than doing a `Future` implementation, I think it would be better to implement our own `ResponseHandler` here.  It should return `incomplete` responses essentially building up a BAOS or something similar until the final chunk, where it can convert that into the snapshot, do the work and finally return a `complete` response.